# Demographic filtering
It is one of the simplest movie recommendation system that offers users the best rated and most popular movies. However it seems simple, we need appropriate formula to measure the rating of a movie to consider it as best among others. Only rating won't give us the proper results cause if a movie has 9.5 rating but number of votes is less than 10, then it biases our result.

So Here we use the popular formula given by IMDB, which is articulated [here](https://help.imdb.com/article/imdb/track-movies-tv/ratings-faq/G67Y87TFYYP6TWAV#calculatetop) clearly. It basically gives the weightage based on number of votes and ratings of that movie. The formula as follows:

$$W = \frac{R*v+C*m}{v+m}$$

Where,  

- $W \to$ Weight of the rating  
- $R \to$ Average rating of the movie
- $v \to$ Number of votes for the movie
- $C \to$ Mean of the average votes of whole data
- $m \to$ Minimum number of votes to be considered

Therefore we need to determine mean of average votes of whole data ($C$) and decide minimum number of votes to be considered ($m$). First lets import the preprocessed data from Preprocess folder and other required modules.

In [1]:
import pandas as pd
import numpy as np
from preprocessing import df
df.columns

Index(['budget', 'genres', 'id', 'keywords', 'original_language',
       'original_title', 'overview', 'popularity', 'production_companies',
       'production_countries', 'release_date', 'revenue', 'runtime',
       'spoken_languages', 'status', 'tagline', 'title', 'vote_average',
       'vote_count', 'cast', 'crew'],
      dtype='object')

Now lets calculate $C$ which is mean of all the votes in the data using vote average column:

In [2]:
C = df["vote_average"].mean()
C

6.094186288810164

For $m$ which is minimum number of votes to be considered. Lets take 80% as our cuttoff, i.e. the movie needs to have more votes than 80% of the movies. We can use the quantile function:

In [3]:
m = df["vote_count"].quantile(0.8)
m

958.4000000000001

This implies a movie must have atleast $\approx$ 958 votes to be considered for based on rating score. So for demographic filtering lets filter out all other votes which have vote count lesser than $m$.

In [4]:
print("Before: ",df.shape)
df = df.copy().loc[df["vote_count"] >= m]
print("After: ", df.shape)

Before:  (4799, 21)
After:  (960, 21)


Now lets create our weightage function which gives rating based on imdb rating formula.

In [5]:
def weight_ratings(row):
    v = row["vote_count"]
    R = row["vote_average"]
    return (R*v+C*m)/(v+m)

Lets create a new column called weightage, for our preferred rating criteria.

In [6]:
df['weightage'] = df.apply(weight_ratings, axis=1)

That's it the work is done, all thats left is sorting the dataframe based on weightage and printing top films.

In [7]:
df.sort_values("weightage", ascending=False, inplace=True)

In [9]:
df[["title","weightage", "vote_average", "vote_count"]].head(50)

title  weightage  \
1881                           The Shawshank Redemption   8.248376   
662                                          Fight Club   8.096165   
3337                                      The Godfather   8.077454   
3232                                       Pulp Fiction   8.074775   
65                                      The Dark Knight   8.044279   
809                                        Forrest Gump   7.972862   
96                                            Inception   7.969319   
95                                         Interstellar   7.937437   
1990                            The Empire Strikes Back   7.904828   
1818                                   Schindler's List   7.900172   
3865                                           Whiplash   7.894419   
329       The Lord of the Rings: The Return of the King   7.886933   
2294                                      Spirited Away   7.859426   
2912                                          Star Wars   7.846469   
1553                                              Se7en   7.814077   
262   The Lord of the Rings: The Fellowship of the Ring   7.810985   
2731                             The Godfather: Part II   7.807948   
690                                      The Green Mile   7.796874   
330               The Lord of the Rings: The Two Towers   7.783725   
77                                           Inside Out   7.757058   
2091                           The Silence of the Lambs   7.744097   
2285                                 Back to the Future   7.740454   
94                              Guardians of the Galaxy   7.738259   
2522                                 The Imitation Game   7.726624   
634                                          The Matrix   7.724570   
3573                                            Memento   7.714477   
494                                       The Lion King   7.711649   
1847                                         GoodFellas   7.706115   
2284                                        The Shining   7.692321   
3057                                 American History X   7.692197   
3719                    One Flew Over the Cuckoo's Nest   7.679494   
298                             The Wolf of Wall Street   7.670142   
1532                           The Grand Budapest Hotel   7.666533   
571                                Inglourious Basterds   7.665756   
1196                                       The Prestige   7.658554   
287                                    Django Unchained   7.652149   
693                                           Gone Girl   7.646248   
3454                                 The Usual Suspects   7.643640   
274                                           Gladiator   7.629469   
628                                 Saving Private Ryan   7.611859   
4300                                     Reservoir Dogs   7.607653   
1490                                 Return of the Jedi   7.592234   
2170                                             Psycho   7.584391   
2760                                               Room   7.582594   
3158                                              Alien   7.581178   
439                                      Shutter Island   7.575876   
2453                                 Dead Poets Society   7.575249   
57                                               WALL·E   7.574640   
351                                        The Departed   7.573294   
88                                           Big Hero 6   7.569525   

      vote_average  vote_count  
1881           8.5        8205  
662            8.3        9413  
3337           8.4        5893  
3232           8.3        8428  
65             8.2       12002  
809            8.2        7927  
96             8.1       13752  
95             8.1       10867  
1990           8.2        5879  
1818           8.3        4329  
3865           8.3        4254  
329            8.1        8064  
2294           8.3        3840  
2912           8.